lib/modules/manager/cargo/artifacts.ts

```diff
@@ -33,24 +36,53 @@ async function cargoUpdate(
   await exec(cmd, execOptions);
 }
 
-export async function updateArtifacts({
-  packageFileName,
-  updatedDeps,
-  newPackageFileContent,
-  config,
-}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
-  logger.debug(`cargo.updateArtifacts(${packageFileName})`);
-
-  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
+async function cargoUpdatePrecise(
+  manifestPath: string,
+  updatedDeps: Upgrade[],
+  constraint: string | undefined
+): Promise<void> {
+  // First update all dependencies that have been bumped in `Cargo.toml`.
+  const cmds = [
+    'cargo update --config net.git-fetch-with-cli=true' +
+      ` --manifest-path ${quote(manifestPath)} --workspace`,
+  ];
 
-  if (
-    !isLockFileMaintenance &&
-    (updatedDeps === undefined || updatedDeps.length < 1)
-  ) {
-    logger.debug('No updated cargo deps - returning null');
-    return null;
+  // Update individual dependencies to their `newVersion`. Necessary when
+  // using the `update-lockfile` rangeStrategy which doesn't touch Cargo.toml.
+  for (const dep of updatedDeps) {
+    cmds.push(
+      `cargo update --config net.git-fetch-with-cli=true` +
+        ` --manifest-path ${quote(manifestPath)}` +
+        ` --package ${dep.packageName!}@${dep.lockedVersion}` +
+        ` --precise ${dep.newVersion}`
+    );
   }
 
+  const execOptions: ExecOptions = {
+    docker: {},
+    toolConstraints: [{ toolName: 'rust', constraint }],
+  };
+
+  await exec(cmds, execOptions);
+}
+
+export async function updateArtifacts(
+  updateArtifact: UpdateArtifact,
+): Promise<UpdateArtifactsResult[] | null> {
+  return await updateArtifactsImpl(updateArtifact);
+}
```

Comment:
Can this refactoring be split into a separate PR to be merged first?

Reply:
Sorry, I'm not sure I have correctly understood your intentions. Isn't this one of the suggestions being made to the original patch https://github.com/renovatebot/renovate/pull/23968#discussion_r1382872136?

Reply:
What I mean is that: you have both added a feature, but also done some refactoring of existing code in order to introduce it. We prefer to split refactoring and feature addition so that it's easier to review. e.g. the first PR is refactor-only, should require little or no changes of tests, etc. Then the feature is added on top, and easier to review on its own.

Reply:
@rarkins Fair enough. I have made https://github.com/renovatebot/renovate/pull/26009 for this, hope this will help with your workflow.